### ISSUE TITLE
feature request: get workspace or work area dimensions, at least for windows

### ISSUE THREAD

**Issue by nerditation** [rust-windowing/winit#2494]
currently you can get `position` as well as `size` of a monitor, but there's no api to get the workspace position and size. see the documentation for SPI_GETWORKAREA [here](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-systemparametersinfoa)

> The work area is the portion of the screen not obscured by the system taskbar or by application desktop toolbars. 

I don't know other platforms, at least on Windows, many programs behave very strangely (e.g. window title bar covered by system taskbar) when the taskbar is at top of the desktop (which I have been using for years), presumably similar situation would happen when taskbar is at left or right. the solution is very simple, calculate your screen coordinate relative to the workspace origin.

on Windows, you get the workspace `RECT` via `GetMonitorInfo` function, which is actually already being caled in current version of `winit` to retrive monitor position and size:
https://github.com/rust-windowing/winit/blob/48b843e42d8b807e8147217b666a732db41db31a/src/platform_impl/windows/monitor.rs#L188

to get the `RECT` of the workspace, you just need to use the `rcWork` field instead of the `rcMonitor` field. I would suggest add two functions: `work_area_position()` and `work_area_size()`, and `MonitorHandleExtWindows` seems to be a reasonal starting place.

what's do you guys think? should I make PR?

in future versions, we might add directly to the `Monitor` type if it turns out to be useful on other platforms too, and fairly easy to imlement.

**Comment by msiglreith** [comment]
Sounds reasonable to me, would be interesting to know how other OS expose/handle this.

**Comment by madsmtm** [comment]
On macOS, you can retrieve the size and position using [`-[NSScreen visibleFrame]`](https://developer.apple.com/documentation/appkit/nsscreen/1388369-visibleframe?language=objc).

Though TBH I don't really see the use-case, can't you just use `WindowBuilder::with_maximized` or `Window::set_maximized`?

**Comment by nerditation** [comment]
> Though TBH I don't really see the use-case, can't you just use `WindowBuilder::with_maximized` or `Window::set_maximized`?

not really. my use case is not to create a window that takes up the full work area (i.e. maximized), but create a transparent overlay window that is positioned relative to the desktop work area. if I use the monitor size instead, the overlay will be positioned on top of the system taskbar when the taskbar is on top of screen.

**Comment by hazelnutcloud** [comment]
Currently trying to build an application that needs this feature. Basically, the application needs to be docked at the right-hand side of the screen with a fixed width and covering the entire height of the work area.

Here's an example of a different application that does exactly this:
![image](https://github.com/user-attachments/assets/bf632cb0-768c-44d9-8537-aff0c9f939e8)

The app above uses electron which already exposes an API to fetch the workArea for a display;[https://www.electronjs.org/docs/latest/api/structures/display](url). Though I couldn't figure out how exactly they implemented this. Would appreciate if this gets added in winit!

**Comment by yezhizhen** [comment]
Servo is using winit for its Desktop release. It needs this feature for https://developer.mozilla.org/en-US/docs/Web/API/Screen/availHeight and `availWidth`.

**Comment by Legend-Master** [comment]
Just in case if someone wants to work on this, and wants some references, we currently implemented this in Tauri like this: https://github.com/tauri-apps/tauri/blob/7f3c989111e007d7eeb5da118421214848e4bfcd/crates/tauri-runtime-wry/src/monitor/

Not entirely sure how to do it on Linux without GTK though

**Comment by kchibisov** [comment]
>Servo is using winit for its Desktop release. It needs this feature for https://developer.mozilla.org/en-US/docs/Web/API/Screen/availHeight and availWidth.

~Just by reading this, it doesn't. `Workarea` is for servo itself, what you've linked is for content inside the servo, so it's always offset from the servo window, not from the desktop space, and you generally in that case assume that your workarea matches monitor resolution, so it's not mandatory, unless you talk about something else.~

Ok, they indeed describe the usecase for a web app to open a window, sounds like that shouldn't be allowed in the first place.

>Not entirely sure how to do it on Linux without GTK though

workarea is present on X11 and non existent on Wayland. GTK 3 has a bug with workarea on Wayland for example breaking pretty much all their popup logic.
